### PR TITLE
Fix incremental build for WPF projects with dotnet.exe

### PR DESCRIPTION
**Pull request by MichaeIDietrich**
This PR fixes incremental builds for WPF projects if `dotnet.exe` is used for building.

The `GetHashCode()` method of strings can produce different hash values between subsequent program runs in .NET Core and cannot be used to persist the `MarkupCompile.cache` file.

A copy of the `GetNonRandomizedHashCode()` method from `String.Comparison.cs` is used instead.

The bug was found as part of #3876.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/wpf?pullRequest=4053) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/wpf?pullRequest=4053) before we can accept your contribution.<br/><br/>:x: MichaeIDietrich [sign now](https://cla.dotnetfoundation.org/dotnet/wpf?pullRequest=4053)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/wpf?pullRequest=4053) it.</sub>

**Comment by ryalanms** [comment]
@MichaeIDietrich: We will discuss this in the WPF triage meeting later today.  Thanks.

File: src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/Tasks/CompilerState.cs [hunk 1]
```diff
@@ -251,6 +251,31 @@ private static string GenerateStringFromFileNames(ITaskItem[] fileItemList)
             return fileNames;
         }
 
+        //
+        // Generates a stable hash code value for strings.
+        // In .NET Core the hash values for the same string can be different between
+        // subsequent program runs and cannot be used for caching here.
+        // source: https://andrewlock.net/why-is-string-gethashcode-different-each-time-i-run-my-program-in-net-core/#a-deterministic-gethashcode-implementation
+        //
+        private static int GetDeterministicHashCode(string str)
```
Comment (fabiant3) [inline src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/Tasks/CompilerState.cs:260]: @MichaeIDietrich - your fix is the correct fix. Can I ask that we instead use the internal String.Comparison method from .NET. Feel free to copy paste: https://source.dot.net/#System.Private.CoreLib/String.Comparison.cs,807

Thank you!

        internal unsafe int GetNonRandomizedHashCode()
        {
            fixed (char* src = &_firstChar)
            {
                Debug.Assert(src[this.Length] == '\0', "src[this.Length] == '\\0'");
                Debug.Assert(((int)src) % 4 == 0, "Managed string should start at 4 bytes boundary");
 
                uint hash1 = (5381 << 16) + 5381;
                uint hash2 = hash1;
 
                uint* ptr = (uint*)src;
                int length = this.Length;
 
                while (length > 2)
                {
                    length -= 4;
                    // Where length is 4n-1 (e.g. 3,7,11,15,19) this additionally consumes the null terminator
                    hash1 = (BitOperations.RotateLeft(hash1, 5) + hash1) ^ ptr[0];
                    hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ ptr[1];
                    ptr += 2;
                }
 
                if (length > 0)
                {
                    // Where length is 4n-3 (e.g. 1,5,9,13,17) this additionally consumes the null terminator
                    hash2 = (BitOperations.RotateLeft(hash2, 5) + hash2) ^ ptr[0];
                }
 
                return (int)(hash1 + (hash2 * 1566083941));
            }
        }

**Review by fabiant3 — CHANGES_REQUESTED** [review]

File: src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/Tasks/CompilerState.cs (hunk 1 above)
Comment (MichaeIDietrich) [reply src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/Tasks/CompilerState.cs]: The `BitOperations` class is not available for the used TFMs. Do you have any preferences for inlining the left rotation code here?

Comment (ryalanms) [reply src/Microsoft.DotNet.Wpf/src/PresentationBuildTasks/MS/Internal/Tasks/CompilerState.cs]: No preference.  Anything is better than what we had before.

**Comment by MichaeIDietrich** [comment]
Can this PR be completed or is there something that is missing?

**Comment by ryalanms** [comment]
Apologies, @MichaeIDietrich.  I didn't see that this PR had been updated.

**Review by ryalanms — APPROVED** [review]

**Comment by ryalanms** [comment]
@fabiant3:  This looks good.  Can you take a look and sign-off?

**Comment by MichaeIDietrich** [comment]
No problem. 😄

**Review by fabiant3 — APPROVED** [review]
Thanks you for the fix @MichaeIDietrich, this should be merged shortly.

**Comment by ryalanms** [comment]
Thanks again, @MichaeIDietrich.

**Pull request merged**